rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+        private final ParametricSummaryBridge summaryBuilder;
+
+        private final AtomicInteger latestVersionCalculateAnalyzer = new AtomicInteger();
+        private final AtomicInteger latestVersionCalculateBuilder = new AtomicInteger();
+
+        @SuppressWarnings("java:S3077")
+        private volatile @MonotonicNonNull CompletableFuture<Optional<VersionedDiagnostics>> calculateAnalyzerLatestResult;
 
         public FileFact(ISourceLocation file) {
             this.file = file;
-            this.summary = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState);
+            this.summaryAnalyzer = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::analyze);
+            this.summaryBuilder = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::build);
         }
 
         public void reloadContributions() {
-            summary.reloadContributions();
+            summaryAnalyzer.reloadContributions();
+            summaryBuilder.reloadContributions();
         }
 
-        private void reportTypeCheckerMessages(List<Diagnostic> messages) {
-            typeCheckerMessages = messages;
-            sendDiagnostics();
+        private VersionedDiagnostics reportDiagnostics(AtomicReference<VersionedDiagnostics> box,
+                VersionedTree tree, List<Diagnostic> messages) {
+
+            var newer = new VersionedDiagnostics(tree, messages);
+            if (VersionedDiagnostics.setIfNewer(box, newer)) {
+                sendDiagnostics();
+            }
+            return newer;
         }
 
-        public void invalidate(boolean isClosing) {
+        private void invalidate(ParametricSummaryBridge summary, boolean isClosing) {
             summary.invalidate(isClosing);
         }
 
-        public void calculate() {
-            summary.calculateSummary();
-            summary.getMessages().thenAccept(this::reportTypeCheckerMessages);
+        public void invalidateAnalyzer(boolean isClosing) {
+            invalidate(summaryAnalyzer, isClosing);
+        }
+
+        public void invalidateBuilder(boolean isClosing) {
+            invalidate(summaryBuilder, isClosing);
+        }
+
+        /**
+         * @param version the version of the file for which summary calculation
+         * is currently requested
+         * @param latestVersion the last version of the file for which summary
+         * calculation was previously requested
+         * @param delay the duration after which the current request for summary
+         * calculation will be granted, unless another request is made in the
+         * meantime (in which case the current request is abandoned)
+         * @param calculation the actual summary calculation
+         * @return a future that provides optional diagnostics that result from
+         * the summary calculation
+         */
+        private CompletableFuture<Optional<VersionedDiagnostics>> calculate(
```

Comment:
should it not be interruptable?

Reply:
`calculate` returns the "debounce future", which is not interruptible (it is created using Java's mechanism of delayed executors). However, to resolve the debounce future, inside, there is an interruptible future that is created in `ParametricSummaryBridge` to perform the actual calculation (after debouncing), which uses the existing interrupt mechanism. (Access to this interruptible future, to interrupt it, goes via a different path than via the "debounce future" returned by `calculate`.)

It should be possible to also handle interrupts in the debounce, but:
  - Right now, interrupts only happen when a new change/save happens. In this case, the current debounce implementation already ensures that obsolete previous `calculate` calls are abandoned. (So interrupting them isn't needed.)
  - The additional bookkeeping will make the code even more complicated.